### PR TITLE
cli(extra-headers): Enable sending additional HTTP Headers via the CLI

### PR DESCRIPTION
**Pull request by rupesh1**
I can across #2746 which mentioned a SO question, that had a link to commit by @fdn - after taking a look it seemed abandoned but functional. So based off their initial commit I've:

- merged in master (had to resolve some conflicts - initial commit was v1.6.0)
- fixed unit tests
- updated some references and function locations to match the current codebase
- added unit tests and basic docs

I've tested the CLI locally and it's working as expected for me.

P.s. I had to run `yarn install-all` a few times for it to run cleanly hence adding yarn-error.log to the gitignore

### CODE REVIEW

File: lighthouse-cli/cli-flags.ts [hunk 1]
```diff
@@ -14,7 +14,8 @@ import {GetValidOutputOptions, OutputMode} from './printer';
 export interface Flags {
   port: number, chromeFlags: string, output: any, outputPath: string, saveArtifacts: boolean,
       saveAssets: boolean, view: boolean, maxWaitForLoad: number, logLevel: string,
-      hostname: string, blockedUrlPatterns: string[], enableErrorReporting: boolean
+      hostname: string, blockedUrlPatterns: string[], enableErrorReporting: boolean,
+      extraHeaders: any
```
Comment (rupesh1) [inline lighthouse-cli/cli-flags.ts]: TODO: Hmm this should probably be of type string

**Comment by rupesh1** [comment]
I signed it!

**Comment by fdn** [comment]
Merge away. Glad this commit is going to see the light of day!

**Comment by johntron** [comment]
@paulirish This PR looks useful for both CLI and programmatic use. If the conflicts are resolved, is there a reason it can't be merged? I know there were other issues where we were discussing using Chrome's _setCookie_.

**Comment by johnboxall** [comment]
This feature would be handy for us: we have conditionals in our code the can be activated/deactivated using HTTP Headers. _Specifically cookies_.

Our current workflow is:

1. Launch an instance of Chrome: `./node_modules/.bin/chrome-debug`. _Note the port number!_
2. In the Chrome instance, set the state you want to test. _For example, navigate to a page and set cookies using the devtools._
3. Clear cached images / files via chrome://settings/clearBrowserData?search=cache
4. In another terminal, run Lighthouse, providing the `--port` argument you noted.

We could script this behaviour using the Chrome Debug Protocol, but ... wouldn't it be sweet if you could just pass an option into 💡 🏠 ?

❤️

File: lighthouse-core/gather/driver.js [hunk 2]
```diff
@@ -947,6 +947,23 @@ class Driver {
...
   }
 
+  /**
+   * @param {!String} jsonHeaders JSON formatted HTTP Header key/value pairs.
+   * @return {!Promise}
+   */
+  setExtraHTTPHeaders(jsonHeaders) {
+    let headers;
+    try {
+      headers = JSON.parse(jsonHeaders);
```
[the remark this answers is not in the document]
Comment (paulirish) [reply lighthouse-core/gather/driver.js]: @rupesh1 holler if you can make this change. if you've moved on, we can pick it up;

**Comment by rupesh1** [comment]
@patrickhulce @paulirish I've had a go at implementing your suggestions (everything passes and runs locally for me node --version 8.9.1, yarn --version 1.3.2), but now bumped in to some typings issue with yargs when the CI runs `yarn unit-cli` (weirdly `yarn type-check` passes on TravisCI):

```
/home/travis/build/GoogleChrome/lighthouse/lighthouse-cli/cli-flags.js:137
      .coerce('extra-headers', /** @param {string} arg */ (arg) => {
       ^
TypeError: y.help(...).version(...).showHelpOnFail(...).usage(...).example(...).example(...).example(...).example(...).example(...).example(...).example(...).example(...).group(...).describe(...).group(...).describe(...).alias(...).group(...).describe(...).boolean(...).choices(...).array(...).string(...).default(...).default(...).default(...).default(...).default(...).default(...).check(...).coerce is not a function
```
I can't reproduce the error locally so any help is appreciated.

**Comment by patrickhulce** [comment]
Thanks for updating @rupesh1!! It looks like you might have a newer version of yargs installed locally in `lighthouse-cli/node_modules` we're quite behind at v3 which doesn't have the `.coerce` option which is why travis is failing

how about we move the parsing logic over to bin.js for now instead?

**Comment by rupesh1** [comment]
@patrickhulce thanks for the tip, CI is looking much better now

**Comment by paulirish** [comment]
I'm not 100% on using a file rather than a string, but neither seems like the clear winner.   Let's do this.

**Comment by paulirish** [comment]
thanks @rupesh1 !

**Comment by asos-rupeshghelani** [comment]
Awesome! Really appreciate everyone's help with this PR 👍

**Comment by rupesh1** [comment]
Awesome! Really appreciate everyone's help with this PR 👍

**Comment by JoelTW** [comment]
This is great. Is there any example of the format required for the json file?

**Comment by patrickhulce** [comment]
@JoelTW it should just be key/value header name/value

```json
{
  "Cookie": "token=12345acdfe",
  "x-secret-header": "it's a secret"
}

**Comment by Pagan-Idel** [comment]
Is there an example we can use these extra headers through programmatical Lighthouse.

**Pull request merged**